### ISSUE TITLE
Disallow using "context" as a property name

### ISSUE THREAD

**Issue by ksol** [emberjs/ember.js#12477]
Hello,

I suggest disallowing (or at least, logging a warning) the use of `context` as a property name.

Granted, it's not very smart to use a name as generic as this one, but it bit me twice in a week, and debugging the first occurence was quite hard and not obvious at all.

For context on what happened: I have a service named `context` that gets injected in all controllers and routes, and in some components.

With a route that had query params (that refreshes the model), having a property named `context` prevented the model from ever being reloaded because an internal property of the same name replaced the service instance I expected, and therefore my method call was throwing an error (which did not appear in the console).

I found it when I copy-pasted the default implementation of "didChangeQueryParams" and putting a `debugger`. Something similar happened this morning with a component, but this time I knew what to look for so it was easier.

Once again, I know it's not smart in the first place to use this name. On the other hand though, nowhere it is said that this name should not or cannot be used, and no obvious warning or error happens when using it.

And since we're there, maybe there's other property names that should be blacklisted, I don't know.

I hope I was clear enough and did not misdiagnosed my issue. Thanks for your time and efforts making emberjs a great piece of software !

**Comment by oriSomething** [comment]
wouldn't it be better to write an ESLint rule instead?

**Comment by ksol** [comment]
Well, I think ESLint rules should be reserved for everything related to the language itself, not the framework internals.

But that's just an opinion. As long as I (the developper) gets notified when I use a property name I should not, I don't really care how it is done.

**Comment by pixelhandler** [comment]
@ksol Seems like a reasonable proposal, but I think the preferece is to use https://github.com/emberjs/ember.js/blob/master/CONTRIBUTING.md#requesting-a-feature 

> Use RFC issues to propose a rough idea

Would you consider opening this suggestion as a RFC issue?

**Comment by ksol** [comment]
I think this request is halfway between feature and bug, and in any case far from being "a rough idea", so a RFC seems a little overkill to me, but if anyone from the core team/issue team tells me to do so, then I'll do it gladly!

**Comment by pixelhandler** [comment]
@ksol it's not a bug, and you have an idea that will help other from creating bugs in their apps, so looks like an 'enhancement'. I think this falls under "requesting a feature". RFC issues are for feature requests that don't need a RFC proposal, simply an issue.

**Comment by corrspt** [comment]
Yeah, just want to pitch in the same thing. I'm using Ember 2.1 and apparently this doesn't happen in Ember 2.2 ([Ember 2.2 - working](https://ember-twiddle.com/6e8580a1d44f4f7b97cb?openFiles=twiddle.json%2Capplication.template.hbs), [Ember 2.1 - not working](https://ember-twiddle.com/74afa9803f0ce4c7cba2?openFiles=twiddle.json%2Capplication.template.hbs)).

But yeah, I also spent a good part of my day trying to debug why a specific component wouldn't re-render on property change, and generally behaved weird. Once I tried renaming the "context" to something else, it worked again.

I have no problem with "system properties" and I know "context" is not a very good name. If there's a list of them that have that issue, I would like to know :+1: In any case, thanks Ember devs for the great work.
